render-wasm/src/render.rs

```diff
@@ -337,58 +381,72 @@ impl RenderState {
+                            debug::render_debug_element(self, element, false);
+                            continue;
+                        } else {
+                            debug::render_debug_element(self, element, true);
+                        }
+                    }
 
-            let layer_rec = skia::canvas::SaveLayerRec::default().paint(&paint);
-            // This is needed so the next non-children shape does not carry this shape's transform
-            self.final_surface.canvas().save_layer(&layer_rec);
+                    let mut paint = skia::Paint::default();
+                    paint.set_blend_mode(element.blend_mode().into());
+                    paint.set_alpha_f(element.opacity());
 
-            self.drawing_surface.canvas().save();
-            if !root_id.is_nil() {
-                self.render_shape(&mut element.clone(), element.clip());
-            } else {
-                self.apply_drawing_to_final_canvas();
-            }
+                    if let Some(image_filter) =
+                        element.image_filter(self.viewbox.zoom * self.options.dpr())
+                    {
+                        paint.set_image_filter(image_filter);
+                    }
 
-            self.drawing_surface.canvas().restore();
+                    let layer_rec = skia::canvas::SaveLayerRec::default().paint(&paint);
+                    self.render_surface.canvas().save_layer(&layer_rec);
 
-            // draw all the children shapes
-            if element.is_recursive() {
-                for id in element.children_ids() {
                     self.drawing_surface.canvas().save();
-                    is_complete = self.render_shape_tree(&id, tree) && is_complete;
+                    if !node_id.is_nil() {
+                        self.render_shape(&mut element.clone(), element.clip());
+                    } else {
+                        self.apply_drawing_to_render_canvas();
+                    }
                     self.drawing_surface.canvas().restore();
+
+                    // Set the node as visited before processing children
+                    self.pending_stack.push((node_id, true));
+
+                    if element.is_recursive() {
+                        for child_id in element.children_ids().iter().rev() {
+                            self.pending_stack.push((*child_id, false));
+                        }
+                    }
+                } else {
+                    self.render_surface.canvas().restore();
                 }
+            } else {
+                return Err(
+                    "Error: Element with root_id {node_id} not found in the tree.".to_string(),
+                );
             }
 
-            self.final_surface.canvas().restore();
-
-            return is_complete;
-        } else {
-            eprintln!("Error: Element with root_id {root_id} not found in the tree.");
-            return false;
+            if get_time() - timestamp > MAX_BLOCKING_TIME_MS {
```

Comment:
I worry the `performance.now()` call we do here might be too expensive to do it after each node. Usually these calls are reserved to just once per frame, whenever posible. A quick workaround could be to batch nodes and only do this call after rendering a specific amount